### DISCUSSION TITLE
Feature Request: Binomial Coefficient Entropy Coding for Inverted Index Postings

### DISCUSSION THREAD

**Discussion by MurageKibicho** [meilisearch/product#334] [Feedback & Feature Proposal]
- Feature Name: Binomial Coefficient Entropy Coding for Inverted Index Postings
- Status: draft
- Start Date: 2021-12-13
- Authors: Murage Kibicho

# Summary
This is a proposal for encoding inverted index integer sequences using a number system based on binomial coefficients.
This number system is chosen because the [minimum number of bits](https://arxiv.org/abs/1908.10598) needed to encode a list of strictly increasing integers
is given by the equation ![ceilinglog2](https://raw.githubusercontent.com/PostingsCompress/PostingsWebsite/main/log2UN130.png)
where ![U](https://raw.githubusercontent.com/PostingsCompress/PostingsWebsite/main/U.png) is largest number in the list and ![n](https://raw.githubusercontent.com/PostingsCompress/PostingsWebsite/main/n.png) is the number of elements in the list.
For example to encode the sequence **1,2,3,4,10,11** we need at least 9 bits: ![11 choose 6](https://raw.githubusercontent.com/PostingsCompress/PostingsWebsite/main/11choose6.png) and
![log2(462) = 8.85 ~ 9 bits](https://raw.githubusercontent.com/PostingsCompress/PostingsWebsite/main/ceil.png).
Binary interpolative coding, one of the best methods for inverted list compression takes at least 20 bits to represent the example sequence. You can confirm this [here](https://github.com/jermp/interpolative_coding).
However, if we represent the same sequence as a sum of binomial coefficients, it takes 10 bits to encode the list, and an extra 3 bits to store the length of the list. The total is 13 bits. You can confirm this [here](https://postingscompress.github.io/PostingsWebsite/#/).

# Motivation
Compression is a solved problem. The best compressors work by changing radixes, or number bases to find the most concise representation of data.
For instance, arithmetic coding is a [generalized change of radix](https://rosettacode.org/wiki/Talk:Arithmetic_coding/As_a_generalized_change_of_radix) for coding
at the [information theoretic entropy bound](https://en.wikipedia.org/wiki/Entropy_(information_theory)). This bound is a measure of redundancy - how many duplicates are in your data. Strictly increasing integer sequences have no duplicates, therefore, cannot be compressed 
according to the information theoretic bound. This means huffman coding and arithmetic coding methods are inefficient with non-repetitive integer sequences.

This rfc proposes the use of the combinatorial number system to encode inverted index integer sequences at the [combinatorial information theoretic entropy bound](https://arxiv.org/abs/1908.10598).
Just like arithmetic coding, this change in number systems allows us to encode integer sequences with the least number of bits.
\
I am a Math major in my junior year and if this RFC succeeds I would love to take a gap year and work on this [library](https://github.com/PostingsCompress/Postings/tree/main/MathLibraries) full time. The library
has sample code for converting between binary and the combinatorial number system using a greedy algorithm, or by generating a lookup table.

# Technical design
## Overview of Combinatorial Number System
Any natural number ![N](https://raw.githubusercontent.com/PostingsCompress/PostingsWebsite/main/N.png) can be uniquely written as a sum of binomial coefficients
using [this](http://math0.wvstateu.edu/~baker/cs405/code/Combinadics.html) greedy algorithm.

1. Find the largest binomial coefficient such that ![akChoosek](https://raw.githubusercontent.com/PostingsCompress/PostingsWebsite/main/akChooseKleqN.png)
2. Subtract to find the residue ![NminusAkChoosek](https://raw.githubusercontent.com/PostingsCompress/PostingsWebsite/main/NMinusBinomial.png)
3. Find the largest binomial coefficient such that ![Repeat](https://raw.githubusercontent.com/PostingsCompress/PostingsWebsite/main/repeat.png)

## **Example:** Find the combinatorial representation of ![n63K4](https://raw.githubusercontent.com/PostingsCompress/PostingsWebsite/main/n63K4.png)
![Convert to Binomial](https://raw.githubusercontent.com/PostingsCompress/PostingsWebsite/main/conversionToBinomials.png)
\
To **reverse** the process, sum your list of binomial coefficients
![Sum](https://raw.githubusercontent.com/PostingsCompress/PostingsWebsite/main/sum130.png)

## Comparison to Binary Interpolative Coding
In the example above, it can be seen that the sequence ![forward sequence](https://raw.githubusercontent.com/PostingsCompress/PostingsWebsite/main/forward.png)
can be encoded in ![for2](https://raw.githubusercontent.com/PostingsCompress/PostingsWebsite/main/for2.png) using the combinatorial number system with an extra 3 bits to store the length of the sequence. This is a total of 9 bits to encode this sequence.
\
Using [this library](https://github.com/jermp/interpolative_coding) it can be confirmed that binary interpolative coding takes between 15 to 23 bits to encode the same sequence.
\
This is a screenshot of the result of binary interpolative coding.
![Screenshot](https://raw.githubusercontent.com/PostingsCompress/PostingsWebsite/main/bic.png) 
\
The combinatorial number systems always encodes integer sequences at the entropy limit.

**Comment by MarinPostma** [comment]
Hey @MurageKibicho !

If I understand your proposal correctly, you propose to change the encoding of the set of documents in the inverted index, so that the set of document ids is represented in the compact format you propose?

Right now, the document ids set are represented in the [RoaringBitmap format](https://roaringbitmap.org/). The key advantage of this format is that, on top of provding decent compression, allows for extremely fast set operation enabled by [SIMD](https://en.wikipedia.org/wiki/SIMD). A key component to changing this encoding would be to have similar time and memory complexity on the set operations as Roaring does. What can you tell us about that?

**Comment by Kerollmops** [comment]
Hey @MurageKibicho!

If I understand correctly your proposal you are developing something similar to [Daniel Lemire's SIMDComp library](https://github.com/lemire/simdcomp)? Which has [a Rust port by Paul Masurel](https://github.com/quickwit-inc/bitpacking). It uses similar principles than the one you showed by using the least amount of bits to represent runs of integers and combines that with SIMD instruction to keep a fast decoding throughput.

As @MarinPostma says we already use the RoaringBitmap format to represent our inverted index integers, as it is easier to work with, it compiles everywhere (even in WASM) and it implements fast set operations already.

Thank you anyway for your proposition it is very kind of you and maybe you can tell us more about the difference with the two libraries I talked about above.

**Reply by MurageKibicho** [reply]
Thank you for your reply! This is the most helpful comment I've received in a while. My library does the same thing as Daniel Lemire's. His library allows for fast encoding and decoding but less compression. My library allows for noticeably slower encoding and decoding but at optimum compression. My intended use case is when you want to store a larger inverted index in memory. His use case is when you want fast inverted index operations. I hope this makes sense. If you would like to talk about this more, please email me at murage.kibicho@yale.edu
